pkg/executor/importer/chunk_process.go

```diff
@@ -48,187 +48,260 @@ var (
+		kvCodec:       kvCodec,
+		groupChecksum: verify.NewKVGroupChecksumWithKeyspace(kvCodec),
 	}
-	p.startOffset = offset
-	return nil
 }
 
-func (p *fileChunkEncoder) encodeLoop(ctx context.Context) error {
-	var err error
-	reachEOF := false
-	prevOffset, currOffset := p.startOffset, p.startOffset
-
-	var encodedBytesCounter, encodedRowsCounter prometheus.Counter
+func (p *chunkEncoder) encodeLoop(ctx context.Context) error {
+	var (
+		encodedBytesCounter, encodedRowsCounter prometheus.Counter
+		readDur, encodeDur                      time.Duration
+		rowCount                                int
+		rowBatch                                = make([]*kv.Pairs, 0, MinDeliverRowCnt)
+		rowBatchByteSize                        uint64
+		currOffset, prevOffset                  int64
+	)
 	metrics, _ := metric.GetCommonMetric(ctx)
 	if metrics != nil {
 		encodedBytesCounter = metrics.BytesCounter.WithLabelValues(metric.StateRestored)
 		// table name doesn't matter here, all those metrics will have task-id label.
 		encodedRowsCounter = metrics.RowsCounter.WithLabelValues(metric.StateRestored, "")
 	}
 
-	for !reachEOF {
-		readPos, _ := p.parser.Pos()
-		if readPos >= p.chunkInfo.Chunk.EndOffset {
-			break
+	recordSendReset := func() error {
+		if len(rowBatch) == 0 {
+			return nil
 		}
-		var readDur, encodeDur time.Duration
-		canDeliver := false
-		rowBatch := make([]deliveredRow, 0, MinDeliverRowCnt)
-		var rowCount, kvSize uint64
-	outLoop:
-		for !canDeliver {
-			readDurStart := time.Now()
-			err = p.parser.ReadRow()
-			readPos, _ = p.parser.Pos()
-			// todo: we can implement a ScannedPos which don't return error, will change it later.
-			currOffset, _ = p.parser.ScannedPos()
-
-			switch errors.Cause(err) {
-			case nil:
-			case io.EOF:
-				reachEOF = true
-				break outLoop
-			default:
-				return common.ErrEncodeKV.Wrap(err).GenWithStackByArgs(p.chunkInfo.GetKey(), currOffset)
-			}
-			readDur += time.Since(readDurStart)
-			encodeDurStart := time.Now()
-			lastRow := p.parser.LastRow()
-			// sql -> kv
-			kvs, encodeErr := p.encoder.Encode(lastRow.Row, lastRow.RowID)
-			encodeDur += time.Since(encodeDurStart)
-
-			if encodeErr != nil {
-				// todo: record and ignore encode error if user set max-errors param
-				err = common.ErrEncodeKV.Wrap(encodeErr).GenWithStackByArgs(p.chunkInfo.GetKey(), currOffset)
-			}
-			p.parser.RecycleRow(lastRow)
-
-			if err != nil {
-				return err
-			}
 
-			rowBatch = append(rowBatch, deliveredRow{kvs: kvs, offset: currOffset})
-			kvSize += kvs.Size()
-			rowCount++
-			// pebble cannot allow > 4.0G kv in one batch.
-			// we will meet pebble panic when import sql file and each kv has the size larger than 4G / maxKvPairsCnt.
-			// so add this check.
-			if kvSize >= MinDeliverBytes || len(rowBatch) >= MinDeliverRowCnt || readPos == p.chunkInfo.Chunk.EndOffset {
-				canDeliver = true
-			}
+		if currOffset >= 0 && metrics != nil {
+			delta := currOffset - prevOffset
```

Comment:
`fileChunkEncoder) init()`(prevOffset) is lost, some part of the file might be accounted multiple times in this metric

Reply:
https://github.com/pingcap/tidb/pull/50871/commits/0fee9cf75b6eb8450080141412d097a7db3f4200